database/migrations/2020_12_12_203831_create_web_authn_tables.php

```diff
@@ -15,34 +13,6 @@ class CreateWebAuthnTables extends Migration
 	 */
 	public function up()
 	{
-		defined('STRING') or define('STRING', 'string');
-
-		Schema::create('web_authn_credentials', function (Blueprint $table) {
-			$table->string('id', 255);
-
-			// Change accordingly for your users table if you need to.
-			$table->unsignedBigInteger('user_id');
-
-			$table->string('name')->nullable();
-			$table->string('type', 16);
-			$table->json('transports');
-			$table->json('attestation_type');
-			$table->json('trust_path');
-			$table->uuid('aaguid');
-			$table->binary('public_key');
-			$table->unsignedInteger('counter')->default(0);
-
-			// This saves the external "ID" that identifies the user. We use UUID default
-			// since it's very straightforward. You can change this for a plain string.
-			// It must be nullable because those old U2F keys do not use user handle.
-			$table->uuid('user_handle')->nullable();
-
-			$table->timestamps();
-			$table->softDeletes(WebAuthnCredential::DELETED_AT);
-			Configs::where('key', '=', 'username')->orWhere('key', '=', 'password')->update(['type_range' => STRING]);
-
-			$table->primary(['id', 'user_id']);
-		});
```

Comment:
Why this change of a past migration?

Reply:
It is adding a table which will be later deleted.

Reply:
fixed [faf44a2](https://github.com/LycheeOrg/Lychee/pull/1469/commits/faf44a2f63086be3902560abe44d2c11a8daf791)

Reply:
> It is adding a table which will be later deleted. This was used by Larapass, but we are no longer using Larapass.

I suspected so, but I am unsure whether I like this kind of approach which plays around with old migrations. I would rather not to.

I get your principal idea. You want to improve the efficiency of the entire migration for users which will start with a clean install in the future. For those users it doesn't make much sense to first create an (empty) table `web_authn_credentials` which is later replaced by another (empty) table `webauthn_credentials`. However, we have this kind of situation in other places, too. The most extreme example of that is [`database/migrations/2021_12_04_181200_refactor_models.php`](https://github.com/LycheeOrg/Lychee/blob/52ea5b8861a7d34b2c5e26d9b5df11053568f774/database/migrations/2021_12_04_181200_refactor_models.php) which basically recreates the whole Lychee DB. For fresh installations it would be much more efficient to simply start with a DB script that creates the Lychee DB as it is supposed to be _now_ instead of running a long sequence of delta migrations.

If I remember correctly, @kamil4 or @d7415 exactly asked that question whether it is possible to have something like that some time ago. Unfortunately, this seems the approach being used by Laravel. Back then we answered the question definitely negative.

If we start to play around with old migrations we may not only consider fresh installations but we must also consider all corner cases like users which are in the middle between two changed migrations and then proceed with a migration which only "sees" one of the changes, etc. pp. I understand that you tried to take care of that by not only changing this migration, but also adopting all other migrations which depend on this one to deal with all possible cases.

But if we decide to go this way and do not only apply it to the current migration, but use it as a general design pattern, then we will soon end up in a hell of endless possibilities and nobody will be able to understand the migrations anymore. It will also become much harder to help people which are stuck in a migration, because there are simply more possibilities how the DB scheme may look like. (At the moment there is always exactly one for each completed migration.)

Hence, I am strongly in favour of _not_ touching old migrations unless we really need to, because the contain a bug for example. For the moment I would keep it like that.

In the long run, when the migration sequence becomes too annoying we should consider to ship out Lychee 5 and make a clean cut. We can then provide a 1st migration script which either creates the DB from scratch for all fresh installations or allows to migrate from the latest 4.x.y version. Moreover, all users who are not on the latest 4.x.y. version would be forced to first migrate to that before they can migrate to Lychee 5. This sort of clean cut would reduce the number of possibilities to two: either it is a fresh install or a migration from the latest Lychee 4.x.y. From there we would start with a new sequence of delta migrations.

I am curious to hear @kamil4 and @d7415 thoughts on that.

PS: Sorry for the long text.

Reply:
I would definitely like to trim some of the migrations. Ideally we could add a migration that laravel would recognise as a shortcut. but there doesn't seem to be a way to do that or similar.

I suppose we could cheat by adding a migration before that checks migration status, and if x, y and z haven't been run yet, does its thing and marks those as complete. Not sure if this would work though and it could easily go wrong.

I agree that changing existing migrations is messy, so I'm inclined to agree with the "wipe the slate clean with v5" approach.